pwiz_tools/Skyline/Util/UtilInstall.cs

```diff
@@ -114,6 +117,94 @@ public void Dispose()
+        /// </summary>
+        public string Filename;
+
+        /// <summary>
+        /// The path to download the file to, and to unzip if Unzip is true.
+        /// </summary>
+        public string InstallPath;
+
+        /// <summary>
+        /// The online location of the file to download.
+        /// </summary>
+        public Uri DownloadUrl;
+
+        /// <summary>
+        /// If true, existing files will be overwritten (including during the unzip step).
+        /// </summary>
+        public bool OverwriteExisting;
+
+        /// <summary>
+        /// If true, the file must be a ZIP file and it will unzipped into InstallPath after downloading.
+        /// </summary>
+        public bool Unzip;
+    }
+
+    public static class JavaDownloadInfo
+    {
+        static string JRE_FILENAME = @"jre-17.0.1";
+        static Uri JRE_URL = new Uri($@"https://pwiz-upload.s3.us-west-2.amazonaws.com/{JRE_FILENAME}.zip");
+        public static string JavaDirectory => Path.Combine(ToolDescriptionHelpers.GetToolsDirectory(), JRE_FILENAME);
+        public static string JavaBinary => Path.Combine(JavaDirectory, JRE_FILENAME, @"bin", @"java.exe");
+
+        public static FileDownloadInfo[] FilesToDownload = { new FileDownloadInfo { Filename = JRE_FILENAME, InstallPath = JavaDirectory, DownloadUrl = JRE_URL, OverwriteExisting = true, Unzip = true } };
+    }
+
+    public static class SimpleFileDownloader
+    {
+        private static readonly string SKYLINE_TOOL_TESTING_MIRROR_URL = @"https://pwiz-upload.s3.us-west-2.amazonaws.com/skyline_tool_testing_mirror";
+
+        public static bool FileAlreadyDownloaded(FileDownloadInfo requiredFile)
+        {
+            return Directory.Exists(requiredFile.InstallPath);
+        }
+
+        public static IEnumerable<FileDownloadInfo> FilesNotAlreadyDownloaded(IEnumerable<FileDownloadInfo> requiredFiles)
+        {
+            return requiredFiles.Where(f => !FileAlreadyDownloaded(f));
+        }
+
+        public static bool DownloadRequiredFiles(IEnumerable<FileDownloadInfo> filesToDownload, ILongWaitBroker waitBroker)
+        {
+            var filesNotAlreadyDownloaded = FilesNotAlreadyDownloaded(filesToDownload).ToList();
+            using (var client = new MultiFileAsynchronousDownloadClient(waitBroker, filesNotAlreadyDownloaded.Count))
+            {
+                foreach (var requiredFile in filesNotAlreadyDownloaded)
+                {
+                    // for functional testing, replace the hostname with the Skyline tool testing mirror path on AWS
+                    // TODO: make this conditional somehow so that once per day or week we run with the original URLs to test that they still work
+                    var downloadUrl = requiredFile.DownloadUrl;
+                    if (Program.FunctionalTest)
+                        downloadUrl = new Uri(Regex.Replace(downloadUrl.OriginalString, ".*/(.*)", $"{SKYLINE_TOOL_TESTING_MIRROR_URL}/$1"));

```

Comment:
The test should be responsible for setting some variable like "TestUrlOverride".
It's not good for this code here to be checking whether "Program.FunctionalTest" is true and then decide where to download from, since it makes it impossible to write a functional test which tests downloading from the real server.

Reply:
My current thinking is that in a near-future PR I'll add some Program level property that will be set on TestRunner to enable testing real URLs. There will be a TC config that runs that once per day or per week (haven't decided). Having that logic here seemed awfully convenient. I can't think of a reason to let tests decide on a case by case basis whether to use the test-only URLs or to use the original ones.